### ISSUE TITLE
Memory Leak When Training YOLOv8

### ISSUE THREAD

**Issue by kasperschnack** [ultralytics/ultralytics#9999]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

I am experiencing a GPU memory leak while training YOLOv8 on a large dataset of around 1 mio. images. The memory usage progressively increases until the system runs out of GPU memory.

### Environment

- **OS**: Ubuntu 22.04
- **Python Version**: 3.9
- Ultralytics YOLOv8.1.38 🚀 Python-3.9.19 torch-2.0.1+cu117 CUDA:0 (NVIDIA GeForce GTX 1080 Ti, 11169MiB)
Setup complete ✅ (8 CPUs, 15.6 GB RAM, 159.9/192.0 GB disk)

### Minimal Reproducible Example

```
import sys

import mlflow
import yaml
from ultralytics import YOLO

params = yaml.safe_load(open("params.yaml"))["train"]

mlflow.pytorch.autolog()
mlflow.set_tracking_uri("https://dagshub.com/kschnack/sortbot.mlflow")

DATA_YAML_PATH = "data.yaml"

with mlflow.start_run():
    # Log the parameters to MLflow
    mlflow.log_params(params)

    # load a pre-trained model
    pre_trained_model = YOLO(params["model_type"])

    # train
    model = pre_trained_model.train(
        data=DATA_YAML_PATH,
        imgsz=params["imgsz"],
        batch=params["batch"],
        epochs=params["epochs"],
        optimizer=params["optimizer"],
        lr0=params["lr0"],
        seed=params["seed"],
        pretrained=params["pretrained"],
        augment=False,
        name=params["name"],
        project=params["project"],
    )
```

### Additional

I have tried reducing the batch size and image size (imgsz), but the issue persists.

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@kasperschnack hello! Thank you for bringing up this issue and providing a detailed explanation, as well as your willingness to help with a PR! 🌟

Based on what you've described, it seems like a challenging issue. Since you've already tried the common approach of decreasing the batch size and image size without success, we might need to explore a couple of alternative steps.

One thing that could potentially help is ensuring you have the latest version of PyTorch and the dependencies of Ultralytics YOLOv8, as updates might have addressed similar issues. If you haven't already, please try updating your environment:
```bash
pip install --upgrade torch torchvision
```

Additionally, integrating a manual garbage collection could be helpful in controlling the GPU memory usage. You might introduce Python's `gc.collect()` at strategic points, especially after significant operations like completing an epoch:
```python
import gc

# Place this at the end of an epoch or after a significant memory allocation
gc.collect()
torch.cuda.empty_cache()  # If using PyTorch to help release unreferenced memory
```

Remember, integrating manual garbage collection should be considered a temporary fix. The root cause of the memory leak should be identified for a permanent solution.

Since you're willing to contribute with a PR, it would be fantastic if you could experiment with these suggestions or any other approach you think might resolve the issue. Once you have a potential solution or more insight into what's causing the leak, please feel free to share your findings through a PR or further discussion in the issue. Your contribution would greatly benefit the community! 

If I can be of further assistance or if you've got more insights while debugging, please don't hesitate to reach out. Thank you for contributing to making Ultralytics YOLOv8 even better! 😊👍

**Comment by kasperschnack** [comment]
Hi @glenn-jocher

Thank you for your reply! I have searched but havent been able to find any documentation for how to train with a loop, which would allow for garbage collecting in between epochs. Its also a little bit hard to read the code so i cannot figure out if the model.train actually returns the trained model or not. Therefore I'm kind of shooting in the blind with this suggestion - do you think it could work?

```
import sys
import yaml
import gc
import torch
import mlflow
from ultralytics import YOLO

# Load parameters
params = yaml.safe_load(open("params.yaml"))["train"]

# MLflow setup
mlflow.pytorch.autolog()
mlflow.set_tracking_uri("https://dagshub.com/kschnack/sortbot.mlflow")

DATA_YAML_PATH = "data.yaml"

# Initialize the model
model = YOLO(params["model_type"])

# Training loop with manual epoch control
with mlflow.start_run():
    mlflow.log_params(params)

    for epoch in range(params["epochs"]):
        print(f"Starting epoch {epoch + 1}/{params['epochs']}")

        # Simulate training for one epoch
        model = model.train(
            data=DATA_YAML_PATH,
            imgsz=params["imgsz"],
            batch=params["batch"],
            epochs=1,
            optimizer=params["optimizer"],
            lr0=params["lr0"],
            seed=params["seed"],
            pretrained=params["pretrained"],
            augment=False,
            name=params["name"],
            project=params["project"],
        )

        # After epoch cleanup
        gc.collect()
        if torch.cuda.is_available():
            torch.cuda.empty_cache()

        # Log additional information if needed
        mlflow.log_metrics({"epoch": epoch + 1})

        print("Finished epoch", epoch + 1)
```

**Comment by kasperschnack** [comment]
The snippet I send of course didnt work as the model.train doesnt return itself. I did a bit more digging and found your code snippet from [here](https://docs.ultralytics.com/modes/train/#tensorboard):

```
import torch
from ultralytics import YOLO
from torch.utils.data import DataLoader

# Load your model and dataset
model = YOLO('yolov8n.yaml').to('cuda')
dataset = ... # Your dataset here
dataloader = DataLoader(dataset, batch_size=..., shuffle=True)

optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

for epoch in range(num_epochs):
    for images, targets in dataloader:
        images = images.to('cuda')
        targets = targets.to('cuda')

        # Forward pass
        loss, outputs = model(images, targets)

        # Backward and optimize
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

    print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item()}')
```

Can't get it to work though unfortunately

**Comment by glenn-jocher** [comment]
Hey there! 👋 

Thanks for diving deep into this and giving it a go with the new code snippet! Custom training loops can be a bit tricky to set up initially, especially when adapting high-level APIs for tasks like this.

From the snippet you've shared and the environment setup, it seems like you're all set with the updated libraries, which is great! If the custom training loop from the documentation didn't work outright, it might be due to some small discrepancies or missing links between the dataset preparation and the model expectations.

The example you found is intended to illustrate the process of manually handling the training loop which includes data loading, forward pass, loss calculation, and the optimize step. If you're encountering issues, it could be beneficial to check:

1. If `dataset = ...` is properly implemented and aligned with the expected input format for the YOLO `model`.
2. Ensure your `targets` are structured correctly for the model’s expectations (i.e., each target might need specific formatting like `[class, x_center, y_center, width, height]` normalized to the image's width and height).

For diagnostics, try simplifying your loop first to handle a single batch and see if you can get it right without errors. Once that works, scaling up with the full `DataLoader` loop should be smoother.

Updating libraries was a good move! Sometimes compatibility issues can cause unexpected errors.

Keep experimenting and feel free to share any specific error messages you're running into. The devil is often in the details with these sorts of implementations. Hang in there, and you’ll get it running! 🚀

**Comment by kasperschnack** [comment]
Training failed again but after digging in the logs I found out it was actually system ram, not GPU ram that was running out:

Apr 12 16:32:07 ubuntu-giant kernel: [11115.728385] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(nul
l),cpuset=user.slice,mems_allowed=0,global_oom,task_memcg=/user.slice/user-1000.slice/session-2.scope
,task=python3.9,pid=2923,uid=1000
Apr 12 16:32:07 ubuntu-giant kernel: [11115.728478] Out of memory: Killed process 2923 (python3.9) to
tal-vm:27379012kB, anon-rss:2912616kB, file-rss:101380kB, shmem-rss:505084kB, UID:1000 pgtables:10120
kB oom_score_adj:0
Apr 12 16:32:07 ubuntu-giant systemd[1]: session-2.scope: A process of this unit has been killed by t
he OOM killer.
Apr 12 16:32:07 ubuntu-giant CRON[4642]: (root) CMD ([ -x /etc/init.d/anacron ] && if [ ! -d /run/sys
temd/system ]; then /usr/sbin/invoke-rc.d anacron start >/dev/null; fi)
Apr 12 16:32:31 ubuntu-giant systemd[1]: idle-shutdown.service: Deactivated successfully.

I restarted the script with reduced image size and batch size. Looking at htop I can see that a lot of ram is already beeing taken up by the train.py script as many processes seems to be spawned at the same time
<img width="1440" alt="image" src="https://github.com/ultralytics/ultralytics/assets/8694416/ee4d5a27-72ba-462f-9d86-78c8e9507ab6">

**Comment by kasperschnack** [comment]
For some reason this thing is very RAM hungry, probably because of the size of the dataset. I have increased my swapfile from a measly 2gb to 32gb and set my cache pressure to 50% (sudo sysctl -w vm.vfs_cache_pressure=50) will try again and see if this helps

**Comment by Burhan-Q** [comment]
@kasperschnack system RAM usage can also be reduced by lowering augmentation hyperparameter values or disabling them entirely.

**Comment by kasperschnack** [comment]
Increasing the swap and updating the packages fixed the issue 🧞‍♀️. Thanks for the suggestion @Burhan-Q, will try doing this if using swap decreases speed significantly!

**Comment by H-Tudor** [comment]
I encountered a similar issue while running inference of a videos folder from CLI on CPU with a NANO model with an up-to-date environment (ultralytics 8.2.2). This is new behavior not encountered before in the same flow.

Below are pictures of the task manager \*BEFORE\* crashing (red line RAM, green line SWAP)
![image](https://github.com/ultralytics/ultralytics/assets/53878487/809cbc01-5a54-43c1-a3d0-6bd8f6d1d23c)
![image](https://github.com/ultralytics/ultralytics/assets/53878487/fa2badb4-5575-4026-b6b2-b2227ed49819)

**Comment by glenn-jocher** [comment]
Hey there! 👋 Thanks for reaching out and providing such detailed observations. It indeed seems like you're experiencing something unusual with the updated environment. 🤔

Given that running inference on a folder of videos is pushing your system's memory to its limits, it might be beneficial to try limiting the number of videos processed concurrently, especially if you're operating in a memory-constrained environment like on CPU with a NANO model.

You could potentially manage the video processing flow more granularly to avoid overpowering your system's resources. Here's a quick Python script snippet demonstrating how you could sequentially process each video in your folder, allowing for some cleanup between each to help manage memory usage:

```python
import os
import gc
import subprocess

# List all video files in your directory
video_folder = 'path/to/your/videos'
videos = [f for f in os.listdir(video_folder) if f.endswith('.mp4')]

for video in videos:
    # Construct the CLI command for inference
    command = f'yolo predict model=yolov8n-nano.pt source={os.path.join(video_folder, video)}'
    
    # Execute the command
    subprocess.run(command, shell=True)
    
    # Optional: Perform cleanup
    gc.collect()

    print(f'Finished processing {video}')
```

This might not directly address the root cause, but could help mitigate the issue by ensuring that only one video is processed at a time, thus lowering overall memory demand.

If the problem persists, please consider sharing a bit more about your setup (like the size and number of videos you're processing) so we can dive deeper. 🧐

**Comment by H-Tudor** [comment]
> If the problem persists, please consider sharing a bit more about your setup (like the size and number of videos you're processing) so we can dive deeper. 🧐

UPDATE: after increasing SWAP to 18GB, if previously it managed to process up to 300 videos, now it processed 646 videos (expected since I doubled available memory with swap) (see first attached image from ubuntu task manager)

In terms of setup / context, I'm dealing with 1746 videos with the specs in the second attached image for a total folder size of 18.2 GB.
The inference params
 - model: fine-tuned yolo v8 nano model (trained on a single class)
 - imgsz=640
 - device=cpu
 - conf=0.5

![Task Manager Screenshot](https://github.com/ultralytics/ultralytics/assets/53878487/492639fd-eabc-45bd-a0eb-337654b41b88)

![Video Specs Screenshot](https://github.com/ultralytics/ultralytics/assets/53878487/efed59ab-9e61-4522-9acb-0a1185e3540b)

**Comment by Burhan-Q** [comment]
@H-Tudor the original post was for training, but I'd guess that since you're using videos, you might be referring to inference? If so, you can use `stream=True` which returns a generator and will reduce RAM usage

```py
from ultralytics import YOLO

model = YOLO("YOLOv8.pt")
source= "path/to/videos"
results = model.predict(source, stream=True)

for result in results:
    ... # process results here
```

**Comment by Burhan-Q** [comment]
@hoang-innomize `stream=True` will lower the overall memory usage since it returns a python generator. This is only useful if iterating the results generator and won't be helpful if yielding the results generator into a list or array.

Python garbage collection occurs automatically and even tho there are modules to manually interact with it, system memory will be freed as needed by python. If you _truly_ want to free memory after calling prediction, the python session should be closed and you can accomplish this using CLI commands via the `subprocess` module (see Glenn's comment above). There are certainly tradeoffs in taking this route, but explicit memory management is not a standard feature of python and I don't think there should be any expectation that you can "force" python to release memory. Using the `subprocess` module to perform inference will close the python session which will release the memory after completing.

**Comment by cozeybozey** [comment]
Doesn't it seem like there is a just a memory leak somewhere in the code? 

I am running into a similar issue where after a training session the used memory is not de-allocated. If I then run another training session it ends up using more memory than it did for the first training session. This can repeated until it eventually runs out of memory. 

I also tried multiple different things with manually deleting the model, gc.collect, clearing the cache, etcetera. But none of those things worked. I think it is wrong to think that this is just a python issue, because if you create any other sort of variable and you delete it, the memory is instantly de-allocated. And we are not talking about a few megabytes here, there are multiple gigabytes of memory that remain allocated even though they shouldn't be. 

I ran multiple memory profilers and there were no regular variables or objects that were still allocating some memory. The only thing I found is that according to memray there were some pytorch processes that were still using up a ton of memory even though the model had already been deleted. So maybe there is a memory leak in how you are using pytorch or in pytorch itself?

Btw I am talking about cpu memory here, since that is my bottleneck, but these issues might also exist for gpu memory.
